src/sqlancer/mysql/oracle/MySQLDQEOracle.java

```diff
@@ -0,0 +1,502 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.beust.jcommander.Strings;
+
+import sqlancer.Randomly;
+import sqlancer.common.oracle.DQEBase;
+import sqlancer.common.oracle.TestOracle;
+import sqlancer.common.query.SQLQueryAdapter;
+import sqlancer.common.query.SQLQueryError;
+import sqlancer.common.query.SQLancerResultSet;
+import sqlancer.common.schema.AbstractRelationalTable;
+import sqlancer.common.schema.AbstractTable;
+import sqlancer.mysql.MySQLErrors;
+import sqlancer.mysql.MySQLGlobalState;
+import sqlancer.mysql.MySQLSchema;
+import sqlancer.mysql.MySQLSchema.MySQLColumn;
+import sqlancer.mysql.MySQLSchema.MySQLTable;
+import sqlancer.mysql.MySQLSchema.MySQLTables;
+import sqlancer.mysql.MySQLVisitor;
+import sqlancer.mysql.ast.MySQLExpression;
+import sqlancer.mysql.gen.MySQLExpressionGenerator;
+
+public class MySQLDQEOracle extends DQEBase<MySQLGlobalState> implements TestOracle<MySQLGlobalState> {
+
+    private final MySQLSchema schema;
+
+    public MySQLDQEOracle(MySQLGlobalState state) {
+        super(state);
+        schema = state.getSchema();
+
+        MySQLErrors.addExpressionErrors(selectExpectedErrors);
+
+        MySQLErrors.addExpressionErrors(updateExpectedErrors);
+        updateExpectedErrors.add("cannot be null");
+        updateExpectedErrors.add("Duplicate entry");
+        updateExpectedErrors.add("The value specified for generated column");
+
+        MySQLErrors.addExpressionErrors(deleteExpectedErrors);
+        deleteExpectedErrors.add("a foreign key constraint fails");
+    }
+
+    public void check() throws SQLException {
+
+        MySQLTables tables = schema.getRandomTableNonEmptyTables();
+        String tableName = tables.getTables().stream().map(AbstractTable::getName).collect(Collectors.joining(","));
+
+        // DQE does not support aggregate functions, windows functions
+        // This method does not generate them, may need some configurations if they can be generated
+        MySQLExpressionGenerator expressionGenerator = new MySQLExpressionGenerator(state).setColumns(tables.getColumns());
+        MySQLExpression whereClause = expressionGenerator.generateExpression();
+
+        // MySQLVisitor is not deterministic, we should keep it only once.
+        // Especially, in MySQLUnaryPostfixOperation and MySQLUnaryPrefixOperation
+        String whereClauseStr = MySQLVisitor.asString(whereClause);
+
+        // Generate a SELECT statement
+        List<String> selectColumns = new ArrayList<>();
+        for (MySQLTable table : tables.getTables()) {
+            selectColumns.add(table.getName() + "." + COLUMN_ROWID);
+        }
+        boolean operateOnSingleTable = tables.getTables().size() == 1;
+        boolean generateOrderBy = false;
+        List<String> orderColumns = new ArrayList<>();
+        boolean generateLimit = false;
+        int limit = 0;
+        if (operateOnSingleTable) {
+            if (Randomly.getBooleanWithSmallProbability()) {
```

Comment:
I think the expression generator already has methods for generating `ORDER BY`s, so we should try to re-use those. Could we perhaps also re-use or factor out components from the random query generator?